examples/timeseries/timeseries_weather_forecasting.py

```diff
@@ -0,0 +1,457 @@
+
+    sns.lineplot(x=df2.index.month, y=df2["T (degC)"], ax=ax5)
+    ax5.set(xlabel="month", ylabel="temperature (C)")
+
+    sns.lineplot(x=df2.index.year, y=df2["T (degC)"], ax=ax6)
+    ax6.set(xlabel="year", ylabel="temperature (C)")
+
+    # row 3 of humidity line graphs
+    fig, (ax7, ax8, ax9) = plt.subplots(1, 3, figsize=(10, 8))
+    sns.lineplot(x=df2.index.hour, y=df2["sh (g/kg)"], ax=ax7)
+    ax7.set(xlabel="hour of day", ylabel="specific humidity")
+
+    sns.lineplot(x=df2.index.month, y=df2["sh (g/kg)"], ax=ax8)
+    ax8.set(xlabel="month", ylabel="specific humidity")
+
+    sns.lineplot(x=df2.index.year, y=df2["sh (g/kg)"], ax=ax9)
+    ax9.set(xlabel="year", ylabel="specific humidity")
+
+    plt.show()
+
+
+show_line_graphs(df)
+
+"""
+# ## Data Preprocessing
+# 
+# Here we are picking 300000 data points for training. The observation is recorded every 10 mins, that means 6 times in an hour. Thus STEP is equal to 6. We are tracking data fom past 720 timestamps(i.e. 720/6=120 hours) and this will be used to predict the temperature after 72 timestamps(i.e. 76/6=12 hours).
+# 
+# Raw data is normalized using a z score formula. Since every feature has values with varying ranges, normalization is done to confine the values in a range of [0,1] before training a neural network.
+# It is done by subtracting the mean and dividing by the standard deviation of each feature
+"""
+
+TRAIN_SPLIT = 300000
+PAST = 720
+FUTURE = 72
+STEP = 6
+LEARNING_RATE = 0.001
+VALIDATION_STEPS = 50
+
+BATCH_SIZE = 256
+BUFFER_SIZE = 10000
+EVALUATION_INTERVAL = 200
+EPOCHS = 10
+
+STEPS_PER_EPOCH = TRAIN_SPLIT / 300
+
+
+def normalize(data, train_split):
+    data_mean = data[:train_split].mean(axis=0)
+    data_std = data[:train_split].std(axis=0)
+    return (data - data_mean) / data_std
+
+
+def preprocess_data(data, train_split=TRAIN_SPLIT, past=PAST, future=FUTURE, step=STEP):
+    X_train, y_train, X_val, y_val = [], [], [], []
+    data = normalize(data, train_split)
+    train_data = data[:train_split]
+    val_data = data[train_split:]
+
+    batch_size = BATCH_SIZE
+
+    sequence_length = int(past / step)
+
+    start = past + future
+    end = start + batch_size
+
+    X_train = tf.keras.preprocessing.timeseries_dataset_from_array(
+        train_data,
+        targets=None,
+        sequence_length=sequence_length,
+        sampling_rate=step,
+        batch_size=batch_size,
+    )
+
+    y_train = data[:, 1]
+    y_train = y_train[start:end]
+
+    batch_size = len(data) - train_split - past - future
+    X_val = tf.keras.preprocessing.timeseries_dataset_from_array(
+        val_data,
+        targets=None,
+        sequence_length=sequence_length,
+        sampling_rate=step,
+        batch_size=batch_size,
+    )
+
+    start = train_split + past + future
+    end = len(data) - past
+
+    y_val = data[:, 1]
+    y_val = y_val[start:]
+
+    X_train = np.stack(list(X_train)[0])
+    X_val = np.stack(list(X_val)[0])
```

Comment:
The output of `timeseries_dataset_from_array` is a Dataset object.

If you call `list()` on it, you area iterating over it an extracting the data for each batch. This is super inefficient. You should pass the dataset object as-is to the model.